Exercise: Data Pre-processing
- Name of activity: Data Pre-processing
- Your name: Dima Mikhaylov
- Your UVA computing ID: agp7dp

In [45]:
# Imports and dependencies 
import pandas as pd
import numpy as np

#### Step 1: load and inspect the dataset 

In [61]:
# Source: https://www.fdic.gov/bank/historical/bank/ 
# Data retrieved on 10/22/2021 using download finctionality
filename = "bfb-data.csv"
df = pd.read_csv(filename, encoding = "ISO-8859-1", engine='python')
df.head(2)

Bank Name, City, State Press Release (PR)  \
0                      Almena State Bank, Almena, KS        PR-119-2020   
1  First City Bank of Florida, Fort Walton Beach, FL        PR-112-2020   

  Closing Date Approx. Asset (Millions) Approx. Deposit (Millions)  \
0    23-Oct-20                  $70.00                     $68.70    
1    16-Oct-20                 $134.70                    $131.40    

                              Acquirer & Transaction  Unnamed: 6  
0     Equity Bank has agreed to assume all deposits.         NaN  
1  United Fidelity Bank, fsb has agreed to assume...         NaN

In [11]:
df.shape

(563, 7)

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 563 entries, 0 to 562
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Bank Name, City, State      561 non-null    object 
 1   Press Release (PR)          561 non-null    object 
 2   Closing Date                561 non-null    object 
 3   Approx. Asset (Millions)    562 non-null    object 
 4   Approx. Deposit (Millions)  561 non-null    object 
 5   Acquirer & Transaction      561 non-null    object 
 6   Unnamed: 6                  0 non-null      float64
dtypes: float64(1), object(6)
memory usage: 30.9+ KB


In [25]:
df.describe()

Unnamed: 6
count         0.0
mean          NaN
std           NaN
min           NaN
25%           NaN
50%           NaN
75%           NaN
max           NaN

In [36]:
df.T.describe

<bound method NDFrame.describe of                                                                        0    \
Bank Name, City, State                       Almena State Bank, Almena, KS   
Closing Date                                                     23-Oct-20   
Approx. Asset (Millions)                                           $70.00    
Approx. Deposit (Millions)                                         $68.70    
Acquirer & Transaction      Equity Bank has agreed to assume all deposits.   

                                                                          1    \
Bank Name, City, State      First City Bank of Florida, Fort Walton Beach, FL   
Closing Date                                                        16-Oct-20   
Approx. Asset (Millions)                                             $134.70    
Approx. Deposit (Millions)                                           $131.40    
Acquirer & Transaction      United Fidelity Bank, fsb has agreed to assume...   

          

In [27]:
df.isnull().values.any()

True

In [28]:
df.isnull().sum().sum()

574

#### Conclusion: this is probably a very messy data set that requires cleaning and preprocessing 

In [29]:
df.columns

Index(['Bank Name, City, State', 'Press Release (PR)', 'Closing Date',
       'Approx. Asset (Millions)', 'Approx. Deposit (Millions)',
       'Acquirer & Transaction', 'Unnamed: 6'],
      dtype='object')

#### First, drop columns that are useless for the analysis

In [62]:
df.drop(columns=['Press Release (PR)', 'Unnamed: 6'], axis=1, inplace=True)
df.head(10)

Bank Name, City, State Closing Date  \
0                      Almena State Bank, Almena, KS    23-Oct-20   
1  First City Bank of Florida, Fort Walton Beach, FL    16-Oct-20   
2                First State Bank, Barboursville, WV     3-Apr-20   
3                    Ericson State Bank, Ericson, NE    14-Feb-20   
4       City National Bank of New Jersey, Newark, NJ     1-Nov-19   
5                          Resolute Bank, Maumee, OH    25-Oct-19   
6                  Louisa Community Bank, Louisa, KY    25-Oct-19   
7                   The Enloe State Bank, Cooper, TX    31-May-19   
8   Washington Federal Bank for Savings, Chicago, IL    15-Dec-17   
9  The Farmers and Merchants State Bank of Argoni...    13-Oct-17   

  Approx. Asset (Millions) Approx. Deposit (Millions)  \
0                  $70.00                     $68.70    
1                 $134.70                    $131.40    
2                 $152.40                    $139.50    
3                  $100.90                    $95.20    
4                 $120.60                    $111.20    
5                  $27.10                     $26.20    
6                  $29.70                     $26.50    
7                  $36.70                     $31.30    
8                 $166.30                    $144.00    
9                  $34.20                     $29.60    

                              Acquirer & Transaction  
0     Equity Bank has agreed to assume all deposits.  
1  United Fidelity Bank, fsb has agreed to assume...  
2   MVB Bank has agreed to assume all deposits ex...  
3  Farmers and Merchants Bank has agreed to assum...  
4  Industrial Bank has agreed to assume all depos...  
5  Buckeye State Bank has agreed to assume all de...  
6  Kentucky Farmers Bank Corporation has agreed t...  
7  Legend Bank, N.A. has agreed to assume the ins...  
8  Royal Savings Bank has agreed to assume the no...  
9     Conway Bank has agreed to assume all deposits.

#### This does help a lot as number or missing values went from 574 down to 9

In [33]:
df.isnull().sum().sum()

9

In [41]:
df_nulls = df[df.isnull().any(axis=1)]
df_nulls

Bank Name, City, State Closing Date Approx. Asset (Millions)  \
561                    NaN          NaN                      NaN   
562                    NaN          NaN             $720,682.08    

    Approx. Deposit (Millions) Acquirer & Transaction  
561                        NaN                    NaN  
562                        NaN                    NaN

Source: https://www.fdic.gov/resources/resolutions/bank-failures/failed-bank-list/

In [53]:
data = pd.read_csv('failed_banks.csv')
data.head()

CERT CHCLASS1       CITYST     COST    FAILDATE    FIN    ID  \
0  21111        N   NEWARK, NJ   2491.0  11/01/2019  10534  4100   
1  58112       NM   LOUISA, KY   4547.0  10/25/2019  10532  4098   
2  58317       SB   MAUMEE, OH   2188.0  10/25/2019  10533  4099   
3  10716       NM   COOPER, TX  21577.0  05/31/2019  10531  4097   
4  30570       SB  CHICAGO, IL  86826.0  12/15/2017  10530  4096   

                                  NAME  QBFASSET  QBFDEP  RESTYPE RESTYPE1  \
0     CITY NATIONAL BANK OF NEW JERSEY    120574  111234  FAILURE       PA   
1                LOUISA COMMUNITY BANK     28163   25174  FAILURE       PA   
2                        RESOLUTE BANK     23292   22885  FAILURE       PA   
3                     ENLOE STATE BANK     36738   31254  FAILURE       PI   
4  WASHINGTON FEDERAL BANK FOR SAVINGS    166345  143964  FAILURE       PI   

  SAVR  
0  DIF  
1  DIF  
2  DIF  
3  DIF  
4  DIF

In [54]:
data.shape

(577, 13)

In [56]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 577 entries, 0 to 576
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   CERT      577 non-null    int64  
 1   CHCLASS1  577 non-null    object 
 2   CITYST    577 non-null    object 
 3   COST      563 non-null    float64
 4   FAILDATE  577 non-null    object 
 5   FIN       577 non-null    int64  
 6   ID        577 non-null    int64  
 7   NAME      577 non-null    object 
 8   QBFASSET  577 non-null    int64  
 9   QBFDEP    577 non-null    int64  
 10  RESTYPE   577 non-null    object 
 11  RESTYPE1  577 non-null    object 
 12  SAVR      577 non-null    object 
dtypes: float64(1), int64(5), object(7)
memory usage: 58.7+ KB


In [57]:
data.describe()

CERT          COST           FIN           ID      QBFASSET  \
count    577.000000  5.630000e+02    577.000000   577.000000  5.770000e+02   
mean   31539.859619  1.302998e+05   9762.890815  3812.000000  6.817803e+06   
std    16458.879911  5.876533e+05   1908.077978   166.709828  8.064436e+07   
min       91.000000  0.000000e+00      0.000000  3524.000000  3.785000e+03   
25%    20115.000000  1.480600e+04  10102.000000  3668.000000  9.086700e+04   
50%    32102.000000  3.833400e+04  10246.000000  3812.000000  2.039550e+05   
75%    35356.000000  9.262400e+04  10390.000000  3956.000000  5.048160e+05   
max    58701.000000  1.207221e+07  10534.000000  4100.000000  1.471631e+09   

             QBFDEP  
count  5.770000e+02  
mean   3.274679e+06  
std    4.172512e+07  
min    5.000000e+02  
25%    7.854900e+04  
50%    1.827940e+05  
75%    4.434810e+05  
max    9.546776e+08

In [59]:
data.isnull().sum().sum()

14

In [58]:
data_missing = data[data.isnull().any(axis=1)]
data_missing

CERT CHCLASS1              CITYST  COST    FAILDATE   FIN    ID  \
505   3510        N       CHARLOTTE, NC   NaN  01/16/2009     0  3586   
507  25178        N   SAN FRANCISCO, CA   NaN  01/16/2009     0  3588   
508  27374       NM  SALT LAKE CITY, UT   NaN  01/16/2009     0  3589   
509  33143       SB      ALEXANDRIA, VA   NaN  01/16/2009     0  3590   
510  33318        N      WILMINGTON, DE   NaN  01/16/2009     0  3591   
511  34571       SB        NEW YORK, NY   NaN  01/16/2009     0  3592   
513  35453        N        PORTLAND, OR   NaN  01/16/2009     0  3594   
514  58032        N      PROVIDENCE, RI   NaN  01/16/2009     0  3595   
518   7213        N       LAS VEGAS, NV   NaN  11/23/2008     0  3578   
519  18708       SB      WILMINGTON, DE   NaN  11/23/2008     0  3579   
520  18923       NM    CENTURY CITY, CA   NaN  11/23/2008     0  3580   
521  23360        N     SIOUX FALLS, SD   NaN  11/23/2008     0  3581   
522  58180        N     SIOUX FALLS, SD   NaN  11/23/2008     0  3582   
561  26652        N      BOCA RATON, FL   NaN  03/01/2002  4654  3539   

                                  NAME    QBFASSET     QBFDEP     RESTYPE  \
505               BANK OF AMERICA N.A.  1471631047  954677580  ASSISTANCE   
507    BANK OF AMERICA CALIFORNIA N.A.    21502223        500  ASSISTANCE   
508             MERRILL LYNCH BANK USA    61809503   53620324  ASSISTANCE   
509               COUNTRYWIDE BANK FSB   117978966   43663251  ASSISTANCE   
510             FIA CARD SERVICES N.A.   159637136    9816599  ASSISTANCE   
511  MERRILL LYNCH BANK & TRUST CO FSB    37986413   28539028  ASSISTANCE   
513        BANK OF AMERICA OREGON N.A.    11526309        500  ASSISTANCE   
514  BANK OF AMERICA RHODE ISLAND N.A.    35410586        500  ASSISTANCE   
518     CITIBANK, NATIONAL ASSOCIATION  1207007000  230042000  ASSISTANCE   
519           CITICORP TRUST BANK, FSB    19599414    7231013  ASSISTANCE   
520             CITIBANK (BANAMEX USA)     1322720     876086  ASSISTANCE   
521      CITIBANK (SOUTH DAKOTA), N.A.    77737957   42356986  ASSISTANCE   
522    DEPARTMENT STORES NATIONAL BANK      374903     300881  ASSISTANCE   
561            NET FIRST NATIONAL BANK       37207      31601     FAILURE   

    RESTYPE1 SAVR  
505      A/A  ***  
507      A/A  ***  
508      A/A  ***  
509      A/A  ***  
510      A/A  ***  
511      A/A  ***  
513      A/A  ***  
514      A/A  ***  
518      A/A  ***  
519      A/A  ***  
520      A/A  ***  
521      A/A  ***  
522      A/A  ***  
561       PI  BIF

#### Four 14 failed banks we have `NaN` for costs. These values can be either with zeros, or overall average, or, even better, average % of costs to assets, as bailing out a bigger bank is more expensive. Imputing with zeros for now.

In [63]:
data['COST'].replace(np.nan, 0, inplace=True)

In [64]:
data.isnull().sum().sum()

0

In [70]:
data['RESTYPE'].describe()

count         577
unique          2
top       FAILURE
freq          564
Name: RESTYPE, dtype: object